crates/bevy_render/src/view/visibility/render_layers.rs

```diff
@@ -72,33 +77,28 @@ impl RenderLayers {
     /// Add the given layer.
     ///
     /// This may be called multiple times to allow an entity to belong
-    /// to multiple rendering layers. The maximum layer is `TOTAL_LAYERS - 1`.
-    ///
-    /// # Panics
-    /// Panics when called with a layer greater than `TOTAL_LAYERS - 1`.
+    /// to multiple rendering layers.
     #[must_use]
-    pub const fn with(mut self, layer: Layer) -> Self {
-        assert!((layer as usize) < Self::TOTAL_LAYERS);
-        self.0 |= 1 << layer;
+    pub fn with(mut self, layer: Layer) -> Self {
+        let (buffer_index, bit) = Self::layer_info(layer);
+        self.extend_buffer(buffer_index + 1);
+        self.0[buffer_index] |= bit;
         self
     }
 
     /// Removes the given rendering layer.
-    ///
-    /// # Panics
-    /// Panics when called with a layer greater than `TOTAL_LAYERS - 1`.
     #[must_use]
-    pub const fn without(mut self, layer: Layer) -> Self {
-        assert!((layer as usize) < Self::TOTAL_LAYERS);
-        self.0 &= !(1 << layer);
+    pub fn without(mut self, layer: Layer) -> Self {
+        let (buffer_index, bit) = Self::layer_info(layer);
+        if buffer_index < self.0.len() {
+            self.0[buffer_index] &= !bit;
+        }
         self
     }
 
     /// Get an iterator of the layers.
-    pub fn iter(&self) -> impl Iterator<Item = Layer> {
-        let total: Layer = std::convert::TryInto::try_into(Self::TOTAL_LAYERS).unwrap();
-        let mask = *self;
-        (0..total).filter(move |g| RenderLayers::layer(*g).intersects(&mask))
+    pub fn iter(&self) -> impl Iterator<Item = Layer> + '_ {
+        self.0.iter().copied().flat_map(Self::iter_layers)
```

Comment:
this doesn't work, it returns values in 0-64 regardless of the array index.

```rust
    let tricky_layers = vec![0, 5, 17, 55, 999, 1025, 1026];
    let layers = RenderLayers::from_layers(&tricky_layers);
    let out = layers.iter().collect::<Vec<_>>();
    assert_eq!(tricky_layers, out);
```
```
assertion `left == right` failed
  left: [0, 5, 17, 55, 999, 1025, 1026]
 right: [0, 5, 17, 55, 39, 1, 2]
```

probably need to pass the base offset into iter_layers

Reply:
```rust
    pub fn iter(&self) -> impl Iterator<Item = Layer> + '_ {
        self.0.iter().copied().zip(0..).flat_map(Self::iter_layers)
    }
```

```rust
    fn iter_layers(buffer_and_offset: (u64, usize)) -> impl Iterator<Item = Layer> + 'static {
        let (mut buffer, mut layer) = buffer_and_offset;
        layer *= 64;
...
```

seems to work

Reply:
Added this as a test case.

Reply:
Extremely confused how adding a `zip` caused this CI lint to fail:

```
error[E0277]: the trait bound `(dyn bevy_reflect::Reflect + 'static): std::marker::Copy` is not satisfied
    --> crates/bevy_render/src/view/visibility/render_layers.rs:101:23
     |
101  |         self.0.iter().copied().zip(0..).flat_map(Self::iter_layers)
     |                       ^^^^^^ the trait `std::marker::Copy` is not implemented for `(dyn bevy_reflect::Reflect + 'static)`
     |
note: required by a bound in `std::iter::Iterator::copied`
    --> /Users/char/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:3479:12
     |
3476 |     fn copied<'a, T: 'a>(self) -> Copied<Self>
     |        ------ required by a bound in this associated function
...
3479 |         T: Copy,
     |            ^^^^ required by this bound in `Iterator::copied`

error[E0599]: `std::iter::Copied<bevy_reflect::ListIter<'_>>` is not an iterator
   --> crates/bevy_render/src/view/visibility/render_layers.rs:101:32
    |
101 |         self.0.iter().copied().zip(0..).flat_map(Self::iter_layers)
    |                                ^^^ `std::iter::Copied<bevy_reflect::ListIter<'_>>` is not an iterator
    |
   ::: /Users/char/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/iter/adapters/copied.rs:21:1
    |
21  | pub struct Copied<I> {
    | -------------------- doesn't satisfy `_: Iterator`
    |
    = note: the following trait bounds were not satisfied:
            `std::iter::Copied<bevy_reflect::ListIter<'_>>: std::iter::Iterator`
            which is required by `&mut std::iter::Copied<bevy_reflect::ListIter<'_>>: std::iter::Iterator`

error[E0369]: no implementation for `dyn bevy_reflect::Reflect & dyn bevy_reflect::Reflect`
   --> crates/bevy_render/src/view/visibility/render_layers.rs:112:29
    |
112 |             if (*self_layer & *other_layer) != 0 {
    |                 ----------- ^ ------------ dyn bevy_reflect::Reflect
    |                 |
    |                 dyn bevy_reflect::Reflect
```

Will have to dig deeper.